added test 8a/8b that creates reads 1D or 2D arrays

guj opened the pull request:
with each rank having a few blocks to write to

guj committed:
added test 5a that creates 1D or 2D arrays, with each rank having a few blocks to write to

guj committed:
fixed tabs

guj committed:
fixed end of line white spaces

guj committed:
type matched

guj committed:
fixed travis error

guj committed:
changed to pass coverage  test

guj committed:
fixed style error

guj committed:
for more coverage

guj committed:
more  coverage

guj committed:
Added multiple steps in the test

guj committed:
indentation change

guj committed:
changed float  to double

guj committed:
changed timer output

guj committed:
outputs

guj committed:
cleaning up

guj committed:
added code to expand  coverage

guj committed:
cleaned up unused tests

guj committed:
fixed tabs

guj committed:
fixed compiler warning from CI

ax3l committed:
Cleanup

ax3l commented:
Thank you! I added some clean-ups and unified the formatting to fit the code base.

Should we maybe name this `8a_benchmark_parallel_write.cpp`?
Please see the `... notes ...` inline :)

guj committed:
addressed comments from Axel

guj committed:
added memory profiler

guj committed:
updated with memory profiler output for linux

guj commented:
> Thank you! I added some clean-ups and unified the formatting to fit the code base.
> 
> Should we maybe name this `8a_benchmark_parallel_write.cpp`?
> Please see the `... notes ...` inline :)

addressed hopefully all thanks.    Feel free to name it after memory issue is settled.

ax3l commented on examples/5a_write_parallel.cpp:
```diff
@@ -0,0 +1,399 @@
+	{
+	  std::cout<<" SIMPLE DEMO OF MEMORY UPHOLDING "<<std::endl;
+	std::vector<double> all(bulk, mpi_rank);
+        auto  p = openPMD::shareRaw(all);
```
this does not work. the `all` variable owns the memory and will destroy it at line 246 when running out of scope. `shareRaw` will then point to invalid memory.

Maybe just create `all` as `std::shared_ptr` just as above?

guj committed:
updated for the record. plan to remove a few things later

guj committed:
cleaned up unnecessary code

guj committed:
using sync mode to write out rather than deferred, to reduce memory consumption

guj committed:
tab fix

guj committed:
end of tab fix

guj committed:
fixed typo

guj committed:
added test3 to have one series per iteration

Commit:
revised test structure...

Commit:
comments, updates, and clean ups

Commit:
fixed styles

Commit:
avoid build error from some platform

Commit:
push macro around adios only test

Commit:
more cleanup

Commit:
code coverage improvement

Commit:
fixed tabs..

ax3l commented on examples/5a_write_parallel.cpp:
```diff
@@ -0,0 +1,644 @@
+      new T[size], []( T * d ) {delete[] d;}
+    };
+
+    for(unsigned long  i = 0ul; i < size; i++ )
+      {
+    E.get()[i] = val;
+      }
```
Can you please fix the alignments? :) Best if you configure your editor to replace tabs with 4 spaces by default.

guj replied:
OK

ax3l commented on docs/source/usage/examples.rst:
```diff
@@ -18,6 +18,7 @@ C++
 - `3a_write_thetaMode_serial.cpp <https://github.com/openPMD/openPMD-api/blob/dev/examples/3a_write_thetaMode_serial.cpp>`_: write an azimuthally decomposed mesh
 - `4_read_parallel.cpp <https://github.com/openPMD/openPMD-api/blob/dev/examples/4_read_parallel.cpp>`_: MPI-parallel mesh read
 - `5_write_parallel.cpp <https://github.com/openPMD/openPMD-api/blob/dev/examples/5_write_parallel.cpp>`_: MPI-parallel mesh write
+- `5a_write_parallel.cpp <https://github.com/openPMD/openPMD-api/blob/dev/examples/5a_write_parallel.cpp>`_: creates 1D or 2D arrays, with each rank having a few blocks to write to
```
Shall we maybe name this `8a_benchmark_write1D2D_parallel`?

guj replied:
OK

Commit:
renamed file name

Commit:
changed output names accordingly

Commit:
updated since we renamed test form 5a to 8a

Commit:
CMakeList.txt updated after file name change from 5a to 8a

Commit:
format correction

Commit:
make this test work with bp too

Commit:
removed ADIOS1 support due to crash at compression

Commit:
cleaning up with spaces

Commit:
if bp backend is adios1, skip it.

Commit:
increase code coverage

guj committed:
Added h5 support

guj committed:
style fix

guj committed:
renamed 8b_write t0 8b_read

guj committed:
added read test

guj committed:
fixed style

guj committed:
fixed spaces

guj committed:
minor fix for compiler issue

Commit:
fixed warning

guj committed:
fixed a bug

guj committed:
h5  collective only works with seg=1

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -1107,7 +1107,9 @@ namespace detail
             bp.param.offset, bp.param.extent, IO, bp.name );
 
-        engine.Put( var, ptr );
+    // using sync mode has the same effect as deferred and less memory consumption
+    // at of now.
```
Ideally this file change goes in a separate pull request, since it's unrelated to the tests and would be better documented that way.

Cosmetic fix:
```suggestion
        // using sync mode has the same effect as deferred and less memory consumption
        // as of ADIOS 2.6.0.
```

ax3l commented on examples/8a_benchmark_write1D2D_parallel.cpp:
```diff
@@ -0,0 +1,682 @@
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <openPMD/openPMD.hpp>
+#include "openPMD/auxiliary/Environment.hpp"
```
```suggestion
#include <openPMD/auxiliary/Environment.hpp>
```

ax3l commented on examples/7_extended_write_serial.cpp:
```diff
@@ -1,5 +1,6 @@
 #include <openPMD/openPMD.hpp>
 
+#include "openPMD/auxiliary/Environment.hpp"
```
```suggestion
#include <openPMD/auxiliary/Environment.hpp>

```

ax3l commented on examples/8a_benchmark_write1D2D_parallel.cpp:
```diff
@@ -0,0 +1,682 @@
+ *
+ * all tests call this functions to store and flush 1D data
+ *
+ * @param series ..... opemPMD-api series
```
please just use spaces instead of `...` (you can align the spaces if you like, but not needed)
the syntax here will be parsed and shown at https://openpmd-api.readthedocs.io/en/0.11.1-alpha/_static/doxyhtml/annotated.html (linked in [manual](https://openpmd-api.readthedocs.io/en/0.11.1-alpha/details/doxygen.html))

ax3l commented on examples/8a_benchmark_write1D2D_parallel.cpp:
```diff
@@ -0,0 +1,682 @@
+#include <fstream>
+
+#if openPMD_HAVE_ADIOS2
+#   include <adios2.h>
```
Hm, if we are using ADIOS2 here directly we should expose it in the build system...

It's ok to quick-hack this, but why do we do this here? Is this for validation or direct comparison?

guj replied:
This should be related with the direct adios2 test. And will remove those and no need for this.

ax3l commented on examples/8b_benchmark_read1D2D_parallel.cpp:
```diff
@@ -0,0 +1,659 @@
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <openPMD/openPMD.hpp>
+#include "openPMD/auxiliary/Environment.hpp"
```
```suggestion
#include <openPMD/auxiliary/Environment.hpp>
```

ax3l commented on examples/8b_benchmark_read1D2D_parallel.cpp:
```diff
@@ -0,0 +1,659 @@
+ *  (looking for ADIOS2 or H5)
+ *
+ */
+std::vector<std::string> getBackends() {
+    std::vector<std::string> res;
+#if openPMD_HAVE_ADIOS2
+    if( auxiliary::getEnvString( "OPENPMD_BP_BACKEND", "NOT_SET" ) != "ADIOS1" )
+        res.emplace_back(".bp");
+#endif
+
+#if openPMD_HAVE_HDF5
+    res.emplace_back(".h5");
+#endif
+    return res;
+}
```
You can use `openPMD::getFileExtensions()` if you update against `dev` (added in #768)

guj replied:
getFileExtensions() is a useful addition. And since I don't want to include json files,  will leave it as is.

ax3l reviewed (commented):
Thank you :sparkles: 

Can you please
- address the merge conflict in `Memory.hpp` (Doxygen problem is fixed in `dev`)
- avoid the changes to `examples/7_extended_write_serial.cpp` (revert unless we need it, but then I would say we make it more general by querying the methods in `Series::` that provide available backends)

guj committed:
removed direct ADIOS2 tests

guj committed:
addressed style comments made by Axel

guj committed:
revert ADIOS2IOHandler.cpp back. Will make a seperate pull request

guj committed:
revert example/7_extended_write_serial.cpp as is (only writes h5 files)

guj committed:
removed eof

guj committed:
removed unused function

guj committed:
changed a comment

ax3l commented:
@guj I let this slip too long, sorry for this.
Can you please solve the merge conflict?

guj commented:
They were resolved. sorry forgot to mark off.

But I will check in again as I added new things in there this week.  So
hold on for a day or two.

- Junmin

On Tue, Oct 20, 2020 at 10:58 PM Axel Huebl <notifications@github.com>
wrote:

> @guj <https://github.com/guj> I let this slip too long, sorry for this.
> Can you please solve the merge conflict?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openPMD/openPMD-api/pull/749#issuecomment-713323785>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABY7OFRO57WA25LPP5N54ULSLZ2APANCNFSM4NRGBMJA>
> .
>

Commit:
added more fields, and particles in the write content
added different read patterns

guj commented:
Too many things changed in the recent month. I am closing this one and have a new pull request
